codec-http2/src/test/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionEncoderTest.java

```diff
@@ -224,37 +339,18 @@ public void headersWriteForUnknownStreamShouldCreateStream() throws Exception {
     public void headersWriteShouldCreateHalfClosedStream() throws Exception {
         int streamId = 5;
         when(stream.id()).thenReturn(streamId);
+        when(stream.state()).thenReturn(IDLE);
         mockFutureAddListener(true);
         when(local.createStream(eq(streamId))).thenReturn(stream);
+        when(writer.writeHeaders(eq(ctx), eq(streamId), eq(EmptyHttp2Headers.INSTANCE), eq(0),
+                eq(DEFAULT_PRIORITY_WEIGHT), eq(false), eq(0), eq(true), eq(promise))).thenReturn(future);
         encoder.writeHeaders(ctx, streamId, EmptyHttp2Headers.INSTANCE, 0, true, promise);
         verify(local).createStream(eq(streamId));
         verify(stream).open(eq(true));
-        verify(writer).writeHeaders(eq(ctx), eq(streamId), eq(EmptyHttp2Headers.INSTANCE), eq(0),
-                eq(DEFAULT_PRIORITY_WEIGHT), eq(false), eq(0), eq(true), eq(promise));
-    }
-
-    @Test
-    public void headersWriteAfterDataShouldWait() throws Exception {
-        final AtomicReference<ChannelFutureListener> listener = new AtomicReference<ChannelFutureListener>();
-        doAnswer(new Answer<Void>() {
-            @Override
-            public Void answer(InvocationOnMock invocation) throws Throwable {
-                listener.set((ChannelFutureListener) invocation.getArguments()[0]);
-                return null;
-            }
-        }).when(future).addListener(any(ChannelFutureListener.class));
-
-        // Indicate that there was a previous data write operation that the headers must wait for.
-        when(remoteFlow.lastFlowControlledFrameSent(any(Http2Stream.class))).thenReturn(future);
-        encoder.writeHeaders(ctx, STREAM_ID, EmptyHttp2Headers.INSTANCE, 0, true, promise);
-        verify(writer, never()).writeHeaders(eq(ctx), eq(STREAM_ID), eq(EmptyHttp2Headers.INSTANCE), eq(0),
-                eq(DEFAULT_PRIORITY_WEIGHT), eq(false), eq(0), eq(true), eq(promise));
-
-        // Now complete the previous data write operation and verify that the headers were written.
-        when(future.isSuccess()).thenReturn(true);
-        listener.get().operationComplete(future);
-        verify(writer).writeHeaders(eq(ctx), eq(STREAM_ID), eq(EmptyHttp2Headers.INSTANCE), eq(0),
-                eq(DEFAULT_PRIORITY_WEIGHT), eq(false), eq(0), eq(true), eq(promise));
+        // Trigger the write and mark the promise successful to trigger listeners
+        payloadCaptor.getValue().write(0);
+        promise.trySuccess();
```

Comment:
@louiscryan - Thanks for your investigation.  I don't want to hold this PR because of this issue, because I don't think the behavior changed.  I set up issue https://github.com/netty/netty/issues/3382 to investigate. PRIORITY can be sent/received an any state so I think we are good there.